### DISCUSSION TITLE
"Package would be ignored" warnings for data subdirectories inside package

### DISCUSSION THREAD

**Discussion by mgorny** [pypa/setuptools#3323] [Q&A] [answered]
We're working on [modernizing setuptools usage in translate-toolkit](https://github.com/translate/translate/pull/4587). As part of that, @nijel moved data files into `translate/share` directory and started using `packages = find:`.

The layout for the package resembles:
```
translate
├── [...]
├── __init__.py
├── share
│   ├── langmodels
│   │   ├── afrikaans.lm
│   │   ├── [...]
│   └── stoplist-en
├── [...]
```

Notably, `translate` and its "normal" subpackages all include `__init__.py` but `share` and `share/langmodels` don't.

When building via `python -m build`, we get the following warnings:

```
/tmp/build-env-s7o1om1t/lib/python3.11/site-packages/setuptools/command/build_py.py:153: SetuptoolsDeprecationWarning:     !!


    ############################
    # Package would be ignored #
    ############################
    Python recognizes 'translate.share' as an importable package, however it is
    included in the distribution as "data".
    This behavior is likely to change in future versions of setuptools (and
    therefore is considered deprecated).

    Please make sure that 'translate.share' is included as a package by using
    setuptools' `packages` configuration field or the proper discovery methods.

    You can read more about "package discovery" and "data files" on setuptools
    documentation page.


!!

  check.warn(importable)
/tmp/build-env-s7o1om1t/lib/python3.11/site-packages/setuptools/command/build_py.py:153: SetuptoolsDeprecationWarning:     !!


    ############################
    # Package would be ignored #
    ############################
    Python recognizes 'translate.share.langmodels' as an importable package, however it is
    included in the distribution as "data".
    This behavior is likely to change in future versions of setuptools (and
    therefore is considered deprecated).

    Please make sure that 'translate.share.langmodels' is included as a package by using
    setuptools' `packages` configuration field or the proper discovery methods.

    You can read more about "package discovery" and "data files" on setuptools
    documentation page.


!!

  check.warn(importable)
```

To be honest, I've dug through setuptools documentation and I've only been able to found two solutions that make the warning disappear:
- either adding `__init__.py` to data directories which seems suboptimal
- or using [automatic discovery](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#automatic-discovery) which is documented as experimental

What is the recommended way to install these files?

**Comment by abravalheri** [comment, marked as answer]
Hi @mgorny, would it be possible for you to use `packages = find_namespace:`?

**Reply by abravalheri** [reply]
Do you have any suggestions about changes in the wording?

The error mention "proper discovery methods": the idea here is that the user can chose between the "Automatic discovery" or the "Custom discovery" mentioned in the [docs](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html).

**Reply by mgorny** [reply]
> Hi @mgorny, would it be possible for you to use `packages = find_namespace:`?

Hmm, I suppose it along with:

```
[options.packages.find]
include =
    translate*
```

works. Thanks! It feels a bit weird but if that's the recommended solution, then why not.

**Reply by mgorny** [reply]
> Do you have any suggestions about changes in the wording?

Perhaps explicitly mentioning that namespaces are the right solution would be a good idea. After all, I'm not entirely sure if you'd actually call a directory that doesn't contain any `.py` files a namespace package.

> The error mention "proper discovery methods": the idea here is that the user can chose between the "Automatic discovery" or the "Custom discovery" mentioned in the [docs](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html).

Well, "automatic discovery" was my first idea but if I read the docs correctly, it's not stable enough to be used yet, isn't it?

**Reply by abravalheri** [reply]
Yeah, it might feel a bit strange at first, but I think in the long term it makes a lot of sense (it might be debatable when a directory is a package and when it is data, but at least this way we align ourselves with what Python can successfully use in an `import` statement).

In the future, this change will help us to fix some errors that currently exist because of this behaviour.

**Reply by emlys** [reply]
@abravalheri just to clarify, you're saying that data directories should be included as namespace packages?

I have a similar project structure, also including a translation data directory:
```
invest/
|- src/
   |- natcap/
      |- invest/
         |- __init__.py
         |- various packages
         |- various modules
         |- internationalization/
            |- locales/
               |- en/
                  |- .gitignore
               |- messages.pot
               |- README.md
```
`natcap` is a namespace package, and all the others are regular packages. `setuptools.find_packages('src')` returns all the expected packages except `natcap`. `setuptools.find_namespace_packages('src')` returns all the expected packages _plus_ `natcap.invest.internationalization, natcap.invest.internationalization.locales, natcap.invest.internationalization.locales.en`. I'm confused why these are identified as namespace packages when they contain no python code at all?

**Reply by abravalheri** [reply]
Hi @emlys, the main reason is that after installing your package you will be able to successfully run the following statement in a REPL:

```python
import natcap.invest.internationalization.locales.en
```

This means that (after the installation) Python does recognize these folders as packages. Here setuptools is only aligning the nomenclature with the de-facto Python implementation.

I think I have previously linked a discussion here about what does it mean to be a package in Python (sorry if I didn't, I am replying on my phone). In that discussion, we can see that Python considers any folder as a package (they will be "importable" as long as their names are valid identifiers). We also reached the conclusion that the differntiation between which folder only contains data files and which folder contains modules is arbitrary (some files ending in `.py` can also be interpreted as data files, not necessarily modules, e.g. template files). Since this interpretation is arbitrary, I think it is only fair to replicate the same behaviour that the interpreter uses in practice.

**Reply by abravalheri** [reply]
This is the link for the discussion: https://github.com/pypa/setuptools/issues/3260

**Reply by emlys** [reply]
Thanks @abravalheri, that's very helpful

**Reply by utkonos** [reply]
I ran into this same warning and after reading what feels like everything on the subject of importing resources in the `setuptools` documentation as well as the Python Packaging Guide I have arrived here. I just want to make sure that I'm crystal clear on the best practice.

I had isolated just this problem. Here is my `setup.cfg`:

```
[metadata]
name = testpackage

[options]
zip_safe = True
include_package_data = True
package_dir=
    =src
packages=find:
python_requires = >=3.10.4

[options.package_data]
testing = templates/*.txt

[options.packages.find]
where = src
```

Here is `test.py`
```
import importlib.resources

def doit():
    data = importlib.resources.read_text('testing.templates', 'file.txt')
    print(data)
```

Here is my package layout:
```
test/
 └─ src/
     └─ testing/
         ├─ __init__.py
         ├─ test.py
         └─ templates/
             └─ file.txt
```

The new, correct, best practice syntax is the following, correct?
```
[options.packages.find]
include =
    testing*
```

Would it be correct to say that the documentation here should be updated:
https://packaging.python.org/en/latest/tutorials/packaging-projects/#configuring-metadata

Also, would it be correct to say that the following documentation should mention this as well?
https://setuptools.pypa.io/en/latest/userguide/datafiles.html

**Reply by emlys** [reply]
I'm thinking this means that now:
- `find_namespace_packages` should be used instead of `find_packages`, because PEP 420 has been accepted and the python interpreter looks for namespace packages by default
- the `package_data` kwarg is obsolete, because any "data" within a package will automatically be included as a sub-package (possibly namespace)
- the `include_package_data` kwarg and corresponding `MANIFEST.in` should only be used to include data located _outside_ the package source directories, because data _inside_ the package source directories will automatically be included as sub-packages

Does that sound right?

**Reply by utkonos** [reply]
I've done some testing and found that the change that I made is not needed. I made the following change, and upload to testpypi then running unit tests on the pip installed package from testpypi passes all tests. It also gets rid of the warning.

Old `setup.cfg`:
```
[metadata]
name = testpackage

[options]
zip_safe = True
include_package_data = True
package_dir =
    = src
packages = find:
python_requires = >=3.10.4

[options.package_data]
testing = templates/*.txt

[options.packages.find]
where = src
```

New `setup.cfg`:
```
[metadata]
name = testpackage

[options]
zip_safe = True
include_package_data = True
package_dir =
    = src
packages = find_namespace:
python_requires = >=3.10.4

[options.package_data]
testing = templates/*.txt

[options.packages.find]
where = src
```

The change is `packages = find:` -> `packages = find_namespace:` and that's it. Is this the correct syntax, or are there any other changes that I should make?

**Reply by abravalheri** [reply]
@utkonos, if you use `find_namespace:` when you have folders without a `__init__.py` file, that should be enough. The `[options.packages.find] include` config is mainly useful for people that don't use the `src`-layout [^1].

@emlys, both `package_data` and `include_package_data` refer to data files (not directories) **inside** the package, so they are not obsolete.
With `include_package_data = False` and no `package_data`, setuptools should ignore files that are not `.py` (please let me know if you have find any anomalies on the behaviour).

[^1]: As a rule of thumb I always use `src`-layout + `find_namespace:` in all my projects because that is the most generic configuration and less error-prone. I would recommend for people that don't want to use the `src`-layout, to use `find_namespace:` + `[options.packages.find] include=<package_name>*` (e.g.  `[options.packages.find] include=mypkg*`). `[options.packages.find] exclude=...` can also be used to ensure **auxiliar/dev-only folder** are ignored.

**Reply by utkonos** [reply]
@abravalheri Thanks for this. Sorry for not understanding it the first few tries.

Should the packaging documentation be adjusted from `packages = find:` to `packages = find_namespace:` in tutorial here:
https://packaging.python.org/en/latest/tutorials/packaging-projects/#configuring-metadata

**Reply by abravalheri** [reply]
As a personal opinion, I don't see why not using `packages = find_namespace:` if the project uses the `src`-layout. However there are mixed feelings in the community about PEP 420 and people have different opinions about tooling supporting implicit namespaces by default.

I don't know if the maintainers of `pypa/packaging.python.org` prefer to leave `packages = find:` as it can be safer for people that are *partially* following the tutorial but are trying to avoid the `src`-layout.

**Reply by abravalheri** [reply]
In the latest version, I added a note to the warning advising about replacing `find:` with `find_namespaces:`

https://github.com/pypa/setuptools/blob/499c468a57d240e5bb450bdb6daedc3e559541dd/setuptools/command/build_py.py#L261-L275

**Reply by abravalheri** [reply]
We are working on documentation updates for setuptools in https://github.com/pypa/setuptools/pull/3335. Later this week I should propose something different for the namespaces.

**Reply by vkottler** [reply]
This seems kind of insane to me. We are _explicitly_ specifying data files to `setuptools` and it is complaining right now.

Why should some kind of "magic" like an opaque package discoverer matter **at all**. `setuptools` is being told exactly what to do.

The deprecation warning is also confusing because it states that these data directories are importable packages, but they don't have `__init__.py` files so why would that be the case?

We want to be able to tell `setuptools` what the package's data files are so it can manifest them. That's currently how it works. Why is this changing?
